transport/src/main/java/io/netty/channel/oio/OioByteStreamChannel.java

```diff
@@ -128,21 +129,44 @@ protected void doWriteFileRegion(FileRegion region) throws Exception {
             outChannel = Channels.newChannel(os);
         }
 
-        long written = 0;
-        for (;;) {
-            long localWritten = region.transferTo(outChannel, written);
-            if (localWritten == -1) {
-                checkEOF(region);
-                return;
+        if (region instanceof SliceableFileRegion) {
+            SliceableFileRegion r = (SliceableFileRegion) region;
+            for (;;) {
+                long localWritten = r.transferBytesTo(outChannel, r.transferableBytes());
+                if (localWritten == -1) {
+                    checkEOF(r);
+                    return;
+                }
+                if (!r.isTransferable()) {
+                    return;
+                }
             }
-            written += localWritten;
-
-            if (written >= region.count()) {
-                return;
+        } else {
+            // TODO: consider remove this branch once we move the methods in SliceableFileRegion
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)